File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid)
    {
        int count = 0;
        
        for(int i = 0; i < grid.size(); i++)
        {
            for(int j = 0; j <grid[0].size(); j++)
            {
                if(grid[i][j] == '1')
                {
                    dfs(i, j, grid);
                    count++;
                }
            }
        }
        
        return count;
    }
    
    void dfs(int x, int y, vector<vector<char>>& grid)
    {
        if(x < 0 || x >= grid.size() || y < 0 || y >= grid[0].size() || grid[x][y] == '0'){
            return;
        }
        
        if(grid[x][y] == '1') grid[x][y] = '0';
        
        dfs(x + 1, y, grid);
        dfs(x - 1, y, grid);
        dfs(x, y + 1, grid);
        dfs(x, y - 1, grid);
    }
};

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        lim = len(grid), len(grid[0])
        
        def dfs(x, y):
            
            
            if x < 0 or x >= lim[0] or y < 0 or y >= lim[1] or grid[x][y] == '0':
                return
            
            
            grid[x][y] = "0"
            
            dfs(x + 1, y)
            dfs(x - 1, y)
            dfs(x, y + 1)
            dfs(x, y - 1)
            
            
        count = 0
        
        for i in range(lim[0]):
            for j in range(lim[1]):
                if grid[i][j] == '1':
                    dfs(i, j)
                    count += 1
                    
        return count